derive/src/pyclass.rs

```diff
@@ -316,11 +318,111 @@ pub fn impl_pyclass(attr: AttributeArgs, item: Item) -> Result<TokenStream2, Dia
     };
 
     let ret = quote! {
-        #item
         impl ::rustpython_vm::pyobject::PyClassDef for #ident {
             const NAME: &'static str = #class_name;
             const DOC: Option<&'static str> = #doc;
         }
     };
     Ok(ret)
 }
+
+pub fn impl_pyclass(attr: AttributeArgs, item: Item) -> Result<TokenStream2, Diagnostic> {
+    let (item, ident, attrs) = match item {
+        Item::Struct(struc) => (quote!(#struc), struc.ident, struc.attrs),
+        Item::Enum(enu) => (quote!(#enu), enu.ident, enu.attrs),
+        other => bail_span!(
+            other,
+            "#[pyclass] can only be on a struct or enum declaration"
+        ),
+    };
+
+    let class_def = generate_class_def(&ident, "pyclass", attr, &attrs)?;
+
+    let ret = quote! {
+        #item
+        #class_def
+    };
+    Ok(ret)
+}
+
+pub fn impl_pystruct_sequence(attr: AttributeArgs, item: Item) -> Result<TokenStream2, Diagnostic> {
+    if let Item::Struct(struc) = item {
+        let class_def = generate_class_def(&struc.ident, "pystruct_sequence", attr, &struc.attrs)?;
+        let mut methods = Vec::new();
+        let mut method_references = Vec::new();
+        let mut field_names = Vec::new();
+        for (i, field) in struc.fields.iter().enumerate() {
+            let idx = Index::from(i);
+            if let Some(ref field_name) = field.ident {
+                let method_name = format!("get_{}", field_name);
+                let method_ident = Ident::new(&method_name, field_name.span());
+
+                let method = quote! {
+                    fn #method_ident(
+                        zelf: &::rustpython_vm::obj::objtuple::PyTuple,
+                        _vm: &::rustpython_vm::vm::VirtualMachine)
+                        -> ::rustpython_vm::pyobject::PyObjectRef {
+                        zelf.fast_getitem(#idx)
+                    }
+                };
+                methods.push(method);
+                let field_name_str = field_name.to_string();
+                // TODO add doc to the generated property
+                let method_reference = quote! {
+                    class.set_str_attr(
+                        #field_name_str,
+                        ::rustpython_vm::obj::objproperty::PropertyBuilder::new(ctx)
+                            .add_getter(Self::#method_ident)
+                            .create(),
```

Comment:
You probably don't need to add it as a method that gets put as an item on the struct, you should be able to generate code like:

```rust
class.set_str_attr(
    "#field_name_str",
    ::rustpython_vm::obj::objproperty::PropertyBuilder::new(ctx)
        .add_getter({
            fn getter(zelf: ..., _vm: ...) -> ... { ... }
            getter
        })
        .create(),
);
```

Reply:
I was able to use a closure here, which is quite nice.